src/main/java/io/pravega/connectors/flink/FlinkPravegaWriter.java

```diff
@@ -531,31 +532,73 @@ public void restoreState(List<UUID> list) throws Exception {
             // we can have more than one transaction to check in case of a scale-in event, or
             // for the reasons discussed in the 'notifyCheckpointComplete()' method.
 
-            if (list != null && list.size() > 0) {
-                for (UUID txnId : list) {
-                    if (txnId != null) {
-                        final Transaction<?> txn = pravegaWriter.getTxn(txnId);
-                        final Transaction.Status status = txn.checkStatus();
-
-                        if (status == Transaction.Status.OPEN) {
-                            // that is the case when a crash happened between when the master committed
-                            // the checkpoint, and the sink could be notified
-                            log.info("{} - committing completed checkpoint transaction {} after task restore",
-                                    name(), txnId);
-
-                            txn.commit();
-
-                            log.debug("{} - committed checkpoint transaction {}", name(), txnId);
-
-                        } else if (status == Transaction.Status.COMMITTED || status == Transaction.Status.COMMITTING) {
-                            // that the common case
-                            log.debug("{} - at restore, transaction {} was already committed", name(), txnId);
+            Exception exception = null;
+            EventStreamWriter<T> restorePravegaWriter = null;
+            ClientFactory restoreClientFactory = null;
+            PravegaState pravegaState = list != null && list.size() > 0 ? list.get(0) : null;
+            if (pravegaState != null) {
```

Comment:
I believe we must process all states in the list, not just the first item in the list.    This implies the numerous writers must be opened.  Java 8 streams has a nice `groupingBy` construct to make it easy to gather all the txns for a given scope/stream pair ([ref](http://www.baeldung.com/java-groupingby-collector)).

Reply:
That's right. You may have more than one state here in a scale-down case, where operator gets the state from multiple previous operators.